crates/sparrow-scheduler/src/worker_pool.rs

```diff
@@ -12,71 +13,39 @@ const DEFAULT_THREAD_COUNT: usize = 8;
 /// Number of slots each thread should have in it's local task queue.
 const LOCAL_QUEUE_SIZE: u16 = 32;
 
-#[derive(Debug)]
 pub struct WorkerPool {
     query_id: String,
     injector: Injector,
-    handles: Vec<std::thread::JoinHandle<error_stack::Result<(), Error>>>,
+    workers: Vec<Worker>,
     /// A vector of the pipelines we created.
     pipelines: Vec<Arc<dyn Pipeline>>,
+    /// Track which pipelines / partitions are still running.
+    pending: Arc<PendingSet>,
 }
 
 impl WorkerPool {
```

Comment:
The name `WorkerPool` seems incorrect. At this point:

1. It creates the workers
2. We add the pipelinesn to it
3. We tell it to start, which creates the threads and runs the pipelines on it.

So, it feels more like `WorkerBuilder` or something like that. Thoughts?

Reply:
I still don't dislike `WorkerPool` and `PlanExecutor`. 

`PlanExecutor` takes the plan, then offers an `execute` method to begin processing. I get your point from the other comment about how it's mainly adding pipelines to the worker pool. But I still like it as the `PlanExecutor` because from the caller's perspective it's the thing I call `execute` on. If it were a builder, I'd expect the ol pattern
```
PlanBuilder builder = new PlanBuilder();
builder.add_pipeline(..);
PlanExecutor executor = builder.build();
executor.execute()
```

As for `WorkerPool`, I like it because it hints to me we're working directly with threads (thread pool).

I do want to look at how `Pipelines` and `Tasks` are created to understand the necessity for the cyclic reference though. That looks potentially hard to debug

Reply:
Heh. I spent a week or two trying different things to avoid the cyclic reference. The problem is that:

1. Pipelines need to know how to schedule their various partitions (so they need a "schedule-token")
2. When the worker receives a task (from the queue), it needs the pipeline (to get the code). Call this "execute-token".
3. It's really helpful if the "schedule-token" and "execute-token" are the same. For instance, we don't want to instances of the same partition running -- so the "combined token" can instead increment a count on schedule, and then check it after execution and say "since this was scheduled during execution, I should requeue it".

The cycle emerges from: Pipeline -> (ScheduleToken == ExecuteToken) -> Pipeline.

Reply:
Ah..yeah looking at the code more, that definitely looks tricky, to say the least.